File: 023주차/보석 쇼핑/woosteelz.py
def solution(gems):
    length = len(set(gems))
    start = 0
    end = 0
    have = {gems[0]: 1}
    ans = [1, len(gems)]

    while start < len(gems):

        if len(have) == length:
            if end - start < ans[1] - ans[0]:
                ans = [start + 1, end + 1]
            if have[gems[start]] == 1:
                del have[gems[start]]
            else:
                have[gems[start]] -= 1
            start += 1

        else:
            end += 1
            if end == len(gems):
                break
            if gems[end] in have:
                have[gems[end]] += 1
            else:
                have[gems[end]] = 1
    print(ans)
    return ans


solution(["DIA", "RUBY", "RUBY", "DIA", "DIA", "EMERALD", "SAPPHIRE", "DIA"])
solution(["AA", "AB", "AC", "AA", "AC"])

File: 025주차/가장 긴 팰린드롬/woosteelz.js
function isPalindrome(s) {
    let start = 0;
    let end = s.length - 1;

    while (start < end) {
        if (s[start] !== s[end]) {
            return false;
        }
        start++;
        end--;
    }
    return true;
}

function solution(s) {
    let len = s.length;

    for (let i = len; i >= 0; i--) {
        for (let j = 0; j <= len - i; j++) {
            if (isPalindrome(s.slice(j, j + i))) {
                return i
            }
        }
    }
    return 1
}

console.log(solution("abacde"))